### PR TITLE
Release 5.4.0

### PR DESCRIPTION
**Pull request by Mriegel**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -312,6 +312,8 @@ Localytics.triggerPlacesNotification(228329);
 
 // Location
 Localytics.setLocationMonitoringEnabled(true);
+Localytics.setLocationMonitoringEnabled(true, true);
```
Comment (abashyam) [inline README.md:315]: Native code doesn't support it but the Cordova plugin does. I am thinking this is calling the native code.

File: README.md [hunk 2]
```diff
@@ -387,14 +389,17 @@ Localytics.setCallToActionListener(
 	function success(result) {
 		// result can be one of:
 		// - {"method": "localyticsShouldDeeplink", "params": {"url": "http://www.google.com", "campaign": {"name": "Campaign Name", ...}}}
-		// - {"method": "localyticsDidOptOut", "params": {"optedOut": true/false, "campaign": {"name": "Campaign Name", ...}
+		// - {"method": "localyticsDidOptOut", "params": {"optedOut": true/false, "campaign": {"name": "Campaign Name", ...}}
```
Comment (abashyam) [inline README.md]: Shouldn't there be a closing brace for the method.  See one for campaign and params.

File: README.md [hunk 3]
```diff
@@ -387,14 +389,17 @@ Localytics.setCallToActionListener(
...
 		// - {"method": "localyticsDidPrivacyOptOut", "params": {"privacyOptedOut": true/false, "campaign": {"name": "Campaign Name", ...}
```
Comment (abashyam) [inline README.md]: If you are cleaning up the braces, this line could use an update as well.

File: README.md [hunk 4]
```diff
@@ -415,7 +420,9 @@ var inAppConfig = {
   /*"dismissButtonImageName": "custom_image", // Must be in app's Bundle */
   "aspectRatio": 0.7,
   "backgroundAlpha": 0.75,
-  "offset": 20
+  "offset": 20,
+  "notchFullScreen": true, // Should the In App message render outside of the safe area (eg. around the Notch on iPHone X)
+  "autoHideHomeScreenIndicator": true // Should the home screen indicator be shown
```
Comment (abashyam) [inline README.md]: Comment should be "Should the home screen indicator be hidden when there is no touch detected."

File: src/ios/CDCTADelegate.m [hunk 5]
```diff
@@ -7,83 +7,122 @@
...
+        [self.commandDelegate sendPluginResult:pluginResult callbackId:self.invokedUrlCommand.callbackId];
+    }
     return YES;
 }
 
+ - (void)requestAlwaysAuthorization:(CLLocationManager *)manager {
+    if (self.monitoringDelegate != nil) {
+        [self.monitoringDelegate requestAlwaysAuthorization:manager];
+    }
+    if ([self canTransmitToJS]) {
```
Comment (abashyam) [inline src/ios/CDCTADelegate.m:98]: Why is this needed in the JS world?

File: src/ios/CDCTADelegate.m [hunk 6]
```diff
@@ -7,83 +7,122 @@
...
+        [pluginResult setKeepCallbackAsBool:YES];
+        [self.commandDelegate sendPluginResult:pluginResult callbackId:self.invokedUrlCommand.callbackId];
+    }
+}
+
+ - (void)requestWhenInUseAuthorization:(CLLocationManager *)manager {
+    if (self.monitoringDelegate != nil) {
+        [self.monitoringDelegate requestWhenInUseAuthorization:manager];
+    }
+    if ([self canTransmitToJS]) {
```
Comment (abashyam) [inline src/ios/CDCTADelegate.m:111]: Why is this needed from a JS context.

File: src/ios/LocalyticsPlugin.m [hunk 7]
```diff
@@ -1191,13 +1222,27 @@ - (void)removeMessagingListener:(CDVInvokedUrlCommand *)command {
 - (void)setLocationMonitoringEnabled:(CDVInvokedUrlCommand *)command {
     [self logInput:@"setLocationMonitoringEnabled" withCommand:command];
     NSNumber *enabled = [command argumentAtIndex:0];
+    NSNumber *persist = [command argumentAtIndex:1];
```
Comment (abashyam) [inline src/ios/LocalyticsPlugin.m:1454]: This might make live reload if it's used with Cordova. Cannot add arguments unless they handle undefined case for backward compatibility.

File: README.md (hunk 1 above)
Comment (Mriegel) [reply README.md:315]: Android does support this method

File: README.md (hunk 2 above)
Comment (Mriegel) [reply README.md]: oh ya - I noticed one was missing, didn't see the second, can fix.

File: src/ios/CDCTADelegate.m (hunk 5 above)
Comment (Mriegel) [reply src/ios/CDCTADelegate.m:98]: This isn't js code.. if you mean why do we need this then it protects us from trying to send events to JS when the user hasn't configured it (and therefore don't have the proper variables set to do so).

File: src/ios/LocalyticsPlugin.m (hunk 7 above)
Comment (Mriegel) [reply src/ios/LocalyticsPlugin.m:1454]: Not sure I understand?

Comment (abashyam) [reply src/ios/LocalyticsPlugin.m:1454]: Like react, Cordova apps can be reloaded on an upgrade. when that occurs any additional argument added to a method should support an undefined value for the additional parameters and the behavior should be the same as the one without the argument.

The notion of live reload of an app exists so that the app can be loaded quickly and a start from scratch is not necessary to reduce app launch times. 
-- I learnt this from our friends from Bloomberg. I didn't know this occurred when I reviewed the first release of Cordova / React.

Comment (Mriegel) [reply src/ios/LocalyticsPlugin.m:1454]: I don't think this is a concern - this does handle undefined.

Our code right after which handles `nil`:
```
if ([persist isKindOfClass:[NSNumber class]]) {
  [Localytics persistLocationMonitoring:persist];
}
```
And cordova's doc on `argumentAtIndex`:
>Returns the argument at the given index. If index >= the number of arguments, returns nil. If the argument at the given index is NSNull, returns nil.

File: src/android/LocalyticsPlugin.java [hunk 8]
```diff
@@ -992,6 +1003,9 @@ public void run() {
         } else if (action.equals("setMessagingListener")) {
             messagingListener = new CDMessagingListener(callbackContext, inAppCampaignCache, pushCampaignCache, placesCampaignCache);
             Localytics.setMessagingListener(messagingListener);
+            PluginResult result = new PluginResult(PluginResult.Status.OK);
+            result.setKeepCallback(true);
+            callbackContext.sendPluginResult(result);
             return true;
         } else if (action.equals("removeMessagingListener")) {
             messagingListener = null;
```
Comment (abashyam) [inline src/android/LocalyticsPlugin.java:1011]: Don't we need a callbackContext.success() here for removeMessagingListener so that the callback that's kept is actually freed.

File: src/android/LocalyticsPlugin.java [hunk 9]
```diff
@@ -1067,6 +1092,9 @@ public void run() {
         } else if (action.equals("setLocationListener")) {
             locationListener = new CDLocationListener(callbackContext);
             Localytics.setLocationListener(locationListener);
+            PluginResult result = new PluginResult(PluginResult.Status.OK);
+            result.setKeepCallback(true);
+            callbackContext.sendPluginResult(result);
             return true;
         } else if (action.equals("removeLocationListener")) {
             locationListener = null;
```
Comment (abashyam) [inline src/android/LocalyticsPlugin.java:1100]: Dont we need callbackContent.success() here to free up the listener that has been kept in removeLocationListener.

File: src/android/LocalyticsPlugin.java [hunk 10]
```diff
@@ -1076,11 +1104,14 @@ public void run() {
...
             ctaListener = new CDCTAListener(callbackContext);
             Localytics.setCallToActionListener(ctaListener);
+            callbackContext.success();
             return true;
         } else if (action.equals("removeCallToActionListener")) {
             ctaListener = null;
             Localytics.setCallToActionListener(null);
-            callbackContext.success();
+            PluginResult result = new PluginResult(PluginResult.Status.OK);
+            result.setKeepCallback(true);
```
Comment (abashyam) [inline src/android/LocalyticsPlugin.java]: I think we need to keep in setCallToActionListener not here. Here we should not keep it.

**Review by abashyam — COMMENTED** [review]
there is 1 setKeepCallback that appears to be misplayed in the removeListener instead of in the set.
The comment for autoHideHomeScreenIndicator may need an update.

File: src/android/LocalyticsPlugin.java (hunk 8 above)
Comment (Mriegel) [reply src/android/LocalyticsPlugin.java:1011]: nope - the `PluginResult(PluginResult.Status.OK)` is inherently a `success()`.  `success()` under the hood basically does this code (minus the keep callback)

File: src/android/LocalyticsPlugin.java (hunk 9 above)
Comment (Mriegel) [reply src/android/LocalyticsPlugin.java:1100]: nope

File: src/android/LocalyticsPlugin.java (hunk 10 above)
Comment (Mriegel) [reply src/android/LocalyticsPlugin.java]: whoops ya nice catch

**Review by abashyam — APPROVED** [review]
